hazelcast/src/main/java/com/hazelcast/map/impl/LazyMapEntry.java

```diff
@@ -0,0 +1,170 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl;
+
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.nio.serialization.SerializationService;
+
+import java.io.Serializable;
+import java.util.Map;
+
+import static com.hazelcast.util.Preconditions.checkNotNull;
+
+/**
+ * A {@link java.util.Map.Entry Map.Entry} implementation which serializes/de-serializes key and value objects on demand.
+ * It is beneficial when you need to prevent unneeded serialization/de-serialization
+ * when creating a {@link java.util.Map.Entry Map.Entry}. Mainly targeted to supply a lazy entry to
+ * {@link com.hazelcast.map.EntryProcessor#process(Map.Entry)} method.
+ * <p/>
+ * <STRONG>Note that this implementation is not synchronized and is not thread-safe.</STRONG>
+ *
+ * @see com.hazelcast.map.impl.operation.EntryOperation#createMapEntry(Data, Object)
+ */
+public class LazyMapEntry implements Map.Entry, Serializable {
+
+    private Object keyObject;
+    private Object valueObject;
+    private transient Data keyData;
+    private transient Data valueData;
+
+    private transient boolean modified;
+    private transient SerializationService serializationService;
+
+    public LazyMapEntry() {
+    }
+
+    public LazyMapEntry(Object key, Object value, SerializationService serializationService) {
+        init(key, value, serializationService);
+    }
+
+    public void init(Object key, Object value, SerializationService serializationService) {
+        checkNotNull(key, "key cannot be null");
+        checkNotNull(serializationService, "SerializationService cannot be null");
+
+        keyData = null;
+        keyObject = null;
+
+        if (key instanceof Data) {
+            this.keyData = (Data) key;
+        } else {
+            this.keyObject = key;
+        }
+
+        valueData = null;
+        valueObject = null;
+
+        if (value instanceof Data) {
+            this.valueData = (Data) value;
+        } else {
+            this.valueObject = value;
+        }
+
+        this.serializationService = serializationService;
+    }
+
+    @Override
+    public Object getKey() {
+        if (keyObject == null && serializationService != null) {
+            keyObject = serializationService.toObject(keyData);
+        }
+        return keyObject;
+    }
+
+
+    @Override
+    public Object getValue() {
+        // It is required to check null serializationService because
+        // after subsequent serialization and deserialization of a LazyMapEntry instance
+        // serializationService will be null.
```

Comment:
I didn't understand the comment here, how `serializationService`will be null after serialization and deserialization?